lib/Sema/CSFix.cpp

```diff
@@ -857,6 +857,49 @@ IgnoreContextualType *IgnoreContextualType::create(ConstraintSystem &cs,
       IgnoreContextualType(cs, resultTy, specifiedTy, locator);
 }
 
+bool RemoveUnnecessaryCoercion::diagnose(Expr *root, bool asNote) const {
+  auto &cs = getConstraintSystem();
+  UnnecessaryCoercionFailure failure(root, cs, getFromType(), getToType(),
+                                     getLocator());
+  return failure.diagnose(asNote);
+}
+
+bool RemoveUnnecessaryCoercion::attempt(ConstraintSystem &cs, Type fromType,
+                                        Type toType,
+                                        ConstraintLocatorBuilder locator) {
+  auto last = locator.last();
+  bool isExplicitCoercion =
+      last && last->is<LocatorPathElt::ExplicitTypeCoercion>();
+  if (!isExplicitCoercion)
+    return false;
+
+  auto expr = cast<CoerceExpr>(locator.getAnchor());
+  auto toTypeRepr = expr->getCastTypeLoc().getTypeRepr();
+
+  // only diagnosing for coercion where the left-side is a DeclRefExpr
+  // or a explicit/implicit coercion e.g. Double(1) as Double
+  if (!isa<ImplicitlyUnwrappedOptionalTypeRepr>(toTypeRepr) &&
+      (isa<DeclRefExpr>(expr->getSubExpr()) ||
+       isa<CoerceExpr>(expr->getSubExpr()))) {
+    // let's not warn if coerced type is not known upfront and
+    // if there's no contextual type in the system at this point
+    // for the parent expr meaning that fromType could be
+    // being inferred from the coercion itself.
+    auto parentExpr = cs.getParentExpr(expr);
```

Comment:
Not sure if this comment express it all, the point is that here since we call attempt when both fromType and toType are equal on `matchTypes`. So if the types are equal, the coercedType is a TypeVar(meaning this equality comes from context), and this contextual type is not coming from the parentExpr, the path left from this contextual equality to come from is the coericon itself. 
Does that make sense @xedin? There is another path or case from where this could false?

Reply:
Something like this - `func foo(_ v: Int) { let _ = { v as Int } }`?

Reply:
That's actually not going to be a type variable but this one would be:

```swift
func foo<T>(_ v: T) {
  let _ = { v as T }
}
```

Reply:
Not sure, but in that case the type of `v` would not be a `TypeVarType` right? because it is already know upfront, so we would not have to check for contextual type I think ... 
But I you right on the other comment maybe the best way is to track the contextual type somehow... I just thought this couldn't be a viable solution since there few set o possibilities for contextual type where we are doing this check (I mean, if types are equal and is not from the context of parent is probably the coercion). 
Do you think we should track the contextual type in the constraint system in a map based on the expression(I think getContextualType(Expr*) kinda does that)? 
Or do you there is another better way to track this? 

Also, thank's for all the help with this as always :)

Reply:
> That's actually not going to be a type variable but this one would be:
> 
> ```swift
> func foo<T>(_ v: T) {
>   let _ = { v as T }
> }
> ```

That makes sense, didn't reload before the previous comment 😅 , but right ... so we would need to have a way to know that the contextual `fromType` is coming from the coercion expr. 
Maybe pass the type and get the expr where it is coming from on the constraint system so we can compara anchor == this expr?

Reply:
I think the most principled approach would be to track location information along with the type associated with type variable e.g. `assignFixedType` would have to operate not on types but on `Binding`(s) instead which would have a "source" information represented by either constraint or at least constraint locator. This would require a substantial refactoring of the constraint system, but at the same time would enable couple of other improvements in diagnostics.

I'll think about this some more to try and figure out easier way...

Reply:
So we would have instead a `TypeVariableBinding` in the system to be able to get location information? 

Another thing is 
> "source" information represented by either constraint

Does this mean we have  "source"  information on the actual constraint? And if it has, what kind of information that we have there would help us in this case?

I've added a couple more test cases in 1baf001d3377a0d187249302d4c724af7270147a, they all passing because I still didn't revert the actual `parentExpr` contextual type check yet, but they should cover all cases we discussed here...

Also, thank's for all the help with this as always :)

Reply:
Take a look at how `determineBestBindings` works, that's the algorithm that supplies the potential bindings for each type variable to try.